common/utils/include/claragenomics/utils/buffer.hpp

```diff
@@ -0,0 +1,202 @@
+        : _size(n)
+        , _capacity(n)
+        , _data(nullptr)
+        , _stream(stream)
+        , _allocator(allocator)
+    {
+        assert(_size >= 0);
+        if (_capacity > 0)
+        {
+            _data = static_cast<value_type*>(
+                _allocator->allocate(_capacity * sizeof(value_type), _stream));
+            CGA_CU_CHECK_ERR(cudaStreamSynchronize(_stream));
+        }
+    }
+
+    // Should this be removed and force it to use allocator for a common interface?
+    buffer(std::shared_ptr<Allocator> allocator, cudaStream_t stream = 0)
+        : buffer(0, allocator, stream)
+    {
+    }
+
+    // move constructor
+    buffer(buffer&& r)
+        : _size(std::exchange(r._size, 0))
+        , _capacity(std::exchange(r._size, 0))
+        , _data(std::exchange(r._data, nullptr))
+        , _stream(std::exchange(r._stream, cudaStream_t(0)))
+        , _allocator(std::exchange(r._allocator, std::shared_ptr<Allocator>(nullptr)))
+    {
+    }
+
+    // move assignment operator
+    buffer& operator=(buffer&& r)
+    {
+        _size      = std::exchange(r._size, 0);
+        _capacity  = std::exchange(r._size, 0);
+        _data      = std::exchange(r._data, nullptr);
+        _stream    = std::exchange(r._stream, cudaStream_t(0));
+        _allocator = std::exchange(r._allocator, std::shared_ptr<Allocator>(nullptr));
+        return *this;
+    }
+
+    ~buffer()
+    {
+        if (nullptr != _data)
+        {
+            _allocator->deallocate(_data, _capacity * sizeof(value_type), _stream);
+        }
+    }
+
+    value_type* data() { return _data; }
+
+    const value_type* data() const { return _data; }
+
+    size_type size() const { return _size; }
+
+    void clear() { _size = 0; }
+
+    iterator begin() { return _data; }
+
+    const_iterator begin() const { return _data; }
+
+    iterator end() { return _data + _size; }
+
+    const_iterator end() const { return _data + _size; }
+
+    void reserve(const size_type new_capacity, cudaStream_t stream)
+    {
+        assert(new_capacity >= 0);
+        set_stream(stream);
+        if (new_capacity > _capacity)
+        {
+            value_type* new_data = static_cast<value_type*>(
+                _allocator->allocate(new_capacity * sizeof(value_type), _stream));
+            if (_size > 0)
+            {
+                CGA_CU_CHECK_ERR(cudaMemcpyAsync(new_data, _data, _size * sizeof(value_type),
+                                                 cudaMemcpyDefault, _stream));
+            }
+            if (nullptr != _data)
+            {
+                _allocator->deallocate(_data, _capacity * sizeof(value_type), _stream);
+            }
+            _data     = new_data;
+            _capacity = new_capacity;
+        }
+    }
+    void reserve(const size_type new_capacity)
+    {
+        reserve(new_capacity, _stream);
+    }
+
+    void resize(const size_type new_size, cudaStream_t stream = 0)
```

Comment:
Having a `reserve(size_type,cudaStream_t)` and a `reserve(size_type)` is a much cleaner solution than my suggestion to have an invalid default argument for the stream... Great!
Could you also add an analogous `resize(size_type)` function?

Reply:
done